README.md
# standoutjs

## Installation

### CDN
Insert this code after jQuery import:
```html
<script src="https://cdn.jsdelivr.net/npm/standoutjs@latest/dist/standout.min.js"></script>
```

### NPM
Install using npm:
```
npm install standoutjs
```

### YARN
Install using yarn:
```
yarn add standoutjs
```

Then load it into your app.js:
```javascript
require('standoutjs');
```

## Basic usage
```javascript
$(yourSelector).standout();
```

## Customizing the options
```javascript
$(yourSelector).standout({
        onlyFirstTime: true,
        top: 0.4,
        bottom: 0.2
});
```

## Enabling the lightbox effect
```javascript
$(yourSelector).standout({
        lightBoxEffect: true,
        top: 0.4,
        bottom: 0.2
});
```
_NOTE: the lightbox effect will override both the `onlyFirstTime` and the `oncePerEvent` options in order to work properly._


## Using the events
```javascript
$(yourSelector).standout({
        top: 0.4,
        bottom: 0.2
}).on("ET", function(event, object) {
    console.log("ENTERING FROM TOP " + obj.idx);
});
```
_NOTE: you can chain all the events as you do for any other jQuery event_


## Events list
- **ET**
    - Fired when the element is ENTERING from the TOP of the defined area
- **EXT**
    - Fired when the element is EXITING from the TOP of the defined area
- **EB**
    - Fired when the element is ENTERING from the BOTTOM of the defined area
- **EXB**
    - Fired when the element is EXITING from the BOTTOM of the defined area
- **C**
    - Fired when the element is at the CENTER of the defined area
- **O**
    - Fired when the element is OVER the TOP of the defined area and inside the viewport
- **U**
    - Fired when the element is UNDER the BOTTOM of the defined area and inside the viewport


## Options
Options are sent to the plugin as an object.

Below you can find the name and definition of all the available options for this plugin.

---
```javascript
onlyFirstTime
```
_Boolean - Default: **false**_

Force the plugin to fire the events only once

---
```javascript
oncePerEvent
```
_Boolean - Default: **true**_

Force the plugin to fire the events only once

---
```javascript
showDemoLayout
```
_Boolean - Default: **false**_

Displays the area of effect of the plugin.

---
```javascript
lightBoxEffect
```
_Boolean - Default: **false**_

Activate the lightbox effect for the elements the plugin is enabled for

---
```javascript
backgroundColor
```
_CSS color value - Default: **#000000**_

NOTE: Option NO longer implemented - Will be removed in future versions of the plugin

---
```javascript
top
```
_Float - Default: **0.3**_

This value defines the top limit of the plugin's area of effect (calculated as percentage of the viewport - starting from the top)

I.E. A value of **0.3** means that the top limit is calculated at 30% of the viewport height, starting from the top.

---
```javascript
bottom
```
_Float - Default: **0.3**_

This value defines the bottom limit of the plugin's area of effect (calculated as percentage of the viewport - starting from the bottom)

I.E. A value of **0.3** means that the bottom limit is calculated at 30% of the viewport height, starting from the bottom.

---
```javascript
overlay: {
    backgroundColor: "#000000",
    opacity: "0",
    width: "100%",
    height: "100%",
    position: "fixed",
    top: "0",
    left: "0",
    zIndex: "9999",
    display: "none"
},
```
_Object - Default: **as defined above**_

The CSS properties of the overlay element used for the lightbox effect.

NOTE: for the best effect is advised to leave all the options with their default value.

---
```javascript
overlayId
```
_String - Default: **overlayStandout**_

The default ID used for the overlay element. You can change this if needed.

---
```javascript
enabled
```
_Boolean - Default: **false**_

NOTE: NOT YET IMPLEMENTED - Will be implemented in future versions of the plugin

Enable or disable the plugin.

---

## Requirements
jQuery >= 1.9.1


## Credits
Basic plugin structure from: https://github.com/alvinpascoe/jquery-boilerplate

Tips and technical support: www.manvitech.it

Moral support: Lois, my love


CHANGELOG.md
# Standout jQuery Plugin

## CHANGE LOG

### Standout 3.1.0 2021-04-30

#### New
* Added page for unit test
* Refactored event detection method

#### Fix
* Fixed bug on previous version

#### Changes
* 

### Standout 3.0.0 2021-04-29

#### New
* Added page for unit test
* Refactored event detection method

#### Fix
* Overlay height and display

#### Changes
* 

### Standout 2.3.5 2021-03-09

#### New

#### Fix
* Overlay height and display

#### Changes

### Standout 2.3.4 2021-03-08

#### New
* Using element original display

#### Fix
* Fixed bug on overlay effect not disappearing
* Fixed width for cloned element
* Removed unusued module export
* Refactoring and module export

#### Changes
* Updated README

### Standout 2.3.3 2021-03-05

#### New
* New plugin version

#### Fix
* Code refactoring
* Added module export for npm package

#### Changes

test/under.js
const st = require('../src/standout');

QUnit.module('st');

QUnit.test('StandoutJS Event: Under', assert => {
    
});

gulpfile.js
/* Inspired by Utlimate Jay Mega Gulpfile */
/* https://github.com/jaysalvat/vegas/blob/master/gulpfile.js */
/* global require:true, process:true */
/* jshint laxbreak:true */

(function() {
    'use strict';

    var pkg        = require('./package.json'),
        del        = require('del'),
        yargs      = require('yargs'),
        exec       = require('child_process').exec,
        fs         = require('fs'),
        dateFormat = require('dateformat'),
        spawn      = require('child_process').spawn,
        gulp       = require('gulp'),
        plugins    = require('gulp-load-plugins')();

    /*
        --type major        : bumps from ie. 2.0.0 -> 3.0.0
        --type minor        : bumps from ie. 2.0.0 -> 2.1.0
        --type patch        : bumps from ie. 2.0.0 -> 2.0.1
        --type prerelease   : bumps from ie. 2.0.0 -> 2.0.0-1
    */
    var bumpVersion = yargs.argv.type || 'patch';
    var now         = new Date();

    var settings = {
        name: 'standoutjs',
        banner: {
            content: [
                '/*!-----------------------------------------------------------------------------',
                ' * <%= pkg.description %>',
                ' * v<%= pkg.version %> - built <%= datetime %>',
                ' * Licensed under the MIT License.',
                ' * https://github.com/DonnieRich/standoutjs',
                ' * ----------------------------------------------------------------------------',
                ' * Copyright (C) 2020-<%= year %> Donato Pasquale Riccio',
                ' * https://designaddicted.eu/',
                ' * --------------------------------------------------------------------------*/',
                ''
            ].join('\n'),
            vars: {
                pkg: pkg,
                datetime: dateFormat(now, 'yyyy-mm-dd'),
                year: dateFormat(now, 'yyyy')
            }
        }
    };

    const getPackageJson = function() {
        return JSON.parse(fs.readFileSync('./package.json'));
    };

    function clean(cb) {
        return del([ './dist' ], cb);
    }

    exports.clean = clean;

    function failIfDirty(cb) {
        return exec('git diff-index HEAD --', function(err, output) {
            if (err) {
                return cb(err);
            }
            if (output) {
                return cb('Repository is dirty');
            }
            return cb();
        });
    }

    function failIfNotMain(cb) {
        exec('git symbolic-ref -q HEAD', function(err, output) {
            if (err) {
                return cb(err);
            }
            if (!/refs\/heads\/main/.test(output)) {
                return cb('Branch is not Main');
            }
            return cb();
        });
    }

    exports.failIfNotMain = failIfNotMain;

    function gitTag(cb) {
        const message = 'v' + getPackageJson().version;

        return exec('git tag ' + message, cb);
    }

    exports.gitTag = gitTag;

    function gitAdd(cb) {
        return exec('git add -A', cb);
    }

    exports.gitAdd = gitAdd;

    function gitCommit(cb) {
        const message = 'Build v' + getPackageJson().version;

        return exec('git commit -m "' + message + '"', cb);
    }

    exports.gitCommit = gulp.series(gitAdd, gitCommit);

    function gitPull(cb) {
        return exec('git pull origin main', function(err, output, code) {
            if (code !== 0) {
                return cb(err + output);
            }
            return cb();
        });
    }

    exports.gitPull = gitPull;

    function gitPush(cb) {
        return exec('git push origin main --tags', function(err, output, code) {
            if (code !== 0) {
                return cb(err + output);
            }
            return cb();
        });
    }

    exports.gitCommit = gulp.series(gitAdd, gitCommit, gitPush);

    function npmPublish(cb) {
        exec('npm publish', function(err, output, code) {
            if (code !== 0) {
                return cb(err + output);
            }
            return cb();
        });
    }

    exports.npmPublish = npmPublish;

    function meta(cb) {
        const metadata = {
                date: dateFormat(now, 'yyyy-mm-dd HH:MM'),
                version: 'v' + getPackageJson().version
            },
            json = JSON.stringify(metadata, null, 4);

        fs.writeFileSync('tmp/metadata.json', json);
        fs.writeFileSync('tmp/metadata.js', '__metadata(' + json + ');');

        return cb();
    }

    function bump() {
        return gulp.src([ 'package.json' ])
            .pipe(plugins.bump(
                /^[a-z]+$/.test(bumpVersion)
                    ? { type: bumpVersion }
                    : { version: bumpVersion }
            ))
            .pipe(gulp.dest('.'));
    }

    exports.npmPublish = npmPublish;

    function year() {
        return gulp.src([ './README.md' ])
            .pipe(plugins.replace(/(Copyright )(\d{4})/g, '$1' + dateFormat(now, 'yyyy')))
            .pipe(gulp.dest('.'));
    }

    exports.year = year;

    function lint() {
        return gulp.src('./src/**.js')
            .pipe(plugins.eslint({}))
            .pipe(plugins.eslint.format());
    }

    exports.lint = lint;

    function copy() {
        return gulp.src([ './src/**/*', '!./src/sass', '!./src/sass/**' ])
            .pipe(gulp.dest('./dist'));
    }

    exports.copy = copy;

    function docs() {
        return gulp.src(['./dist/**.js'])
            .pipe(gulp.dest('./docs/demo/js'));
    }

    exports.docs = docs;

    function uglify() {
        return gulp.src('./dist/**/!(*.min.js).js')
            .pipe(plugins.rename({ suffix: '.min' }))
            .pipe(plugins.sourcemaps.init())
            .pipe(plugins.terser({
                compress: {},
                keep_fnames: true,
                output: {
                    comments: /^!/
                }
            }))
            .on('error', function(err) { console.log(err) })
            .pipe(plugins.sourcemaps.write('.'))
            .pipe(gulp.dest('./dist/'));
    }

    exports.copy = copy;

    function header() {
        settings.banner.vars.pkg = getPackageJson();

        return gulp.src('./dist/*.js')
            .pipe(plugins.header(settings.banner.content, settings.banner.vars ))
            .pipe(gulp.dest('./dist/'));
    }

    exports.header = header;

    function changelog(cb) {
        var filename  = 'CHANGELOG.md',
            version   = getPackageJson().version,
            date      = dateFormat(now, 'yyyy-mm-dd'),
            changelog = fs.readFileSync(filename).toString(),
            lastDate =  dateFormat(fs.statSync(filename).mtime, 'yyyy-mm-dd');

        exec('git log --since="' + lastDate + ' 00:00:00" --oneline --pretty=format:"%s"', function(err, stdout) {
            if (err) {
                return cb(err);
            }

            if (!stdout) {
                return cb();
            }

            let prefixes = ['(new)', '(fix)', '(chg)'];
            let changes = {};

            prefixes.forEach(p => {
                changes[p] = stdout.split('\n').filter( line => {
                    return line.indexOf(p) !== -1;
                }).join('\n').replace(/ *\([^)]*\) */g,'').replace(/\n/g, '\n* ');
            });

            const updates = [
                '### Standout ' + version + ' ' + date,
                '',
                '#### New',
                '* ' + changes['(new)'],
                '',
                '#### Fix',
                '* ' + changes['(fix)'],
                '',
                '#### Changes',
                '* ' + changes['(chg)']
            ].join('\n');

            changelog = changelog.replace(/(## CHANGE LOG)/, '$1\n\n' + updates);

            fs.writeFile(filename, changelog, function(err) {
                return cb();
            });
        });
    }

    exports.changelog = changelog;

    function watch() {
        return gulp.watch("./src/**/*", build);
    }

    exports.watch = watch;
    exports.default = watch;

    const build = gulp.series(
        lint,
        clean,
        copy,
        header,
        uglify,
        docs
    );

    exports.build = build;

    const release = gulp.series(
        failIfNotMain, 
        failIfDirty,
        gitPull,
        bump,
        changelog,
        year,
        clean,
        copy,
        header,
        uglify,
        docs,
        gitAdd,
        gitCommit,
        gitTag,
        gitPush,
        npmPublish
    );

    exports.release = release;
})();

/*

NPM Installation
----------------

npm install --save-dev del
npm install --save-dev yargs
npm install --save-dev exec
npm install --save-dev gulp-eslint
npm install --save-dev eslint-config-airbnb-base eslint-plugin-import
npm install --save-dev gulp
npm install --save-dev gulp-load-plugins
npm install --save-dev gulp-bump
npm install --save-dev gulp-header
npm install --save-dev gulp-autoprefixer
npm install --save-dev gulp-terser
npm install --save-dev gulp-sourcemaps
npm install --save-dev gulp-jshint
npm install --save-dev gulp-zip
npm install --save-dev gulp-rename
npm install --save-dev gulp-replace

*/

src/standout.js
(function(jQuery, window, document, undefined) {

    "use strict";

    const pluginName = "standout";
    // Demo layout
    let demoInit = false;

    const allStandout = [];

    const defaults = {
        // Fire the function linked to the event just the first time and not at
        // every subsequent scroll (it will still be fired if the last event is different from the current one)
        onlyFirstTime: false,
        oncePerEvent: true,
        showDemoLayout: false,
        lightBoxEffect: false,
        backgroundColor: "#000000",
        top: 0.3,
        bottom: 0.3,
        overlay: {
            backgroundColor: "#000000",
            opacity: "0",
            width: "100%",
            height: "100vh",
            position: "fixed",
            top: "0",
            left: "0",
            zIndex: "9999",
            display: "none"
        },
        overlayId: "overlayStandout",
        enabled: true,
        window,
        body: "body"
    };

    const lightboxOptions = {
        onlyFirstTime: false,
        oncePerEvent: false
    };

    const objProps = {
        initialized: false,
        originalTop: 0,
        originalLeft: 0,
        elementWidth: 0,
        elementHeight: 0,
        elementTop: 0,
        elementCenter: 0,
        elementBottom: 0,
        elementTopPosition: 0,
        elementCenterPosition: 0,
        elementBottomPosition: 0,
        viewportHeight: 0,
        viewportTop: 0,
        viewportBottom: 0,
        lastViewportTop: 0,
        lastEvent: "",
        currentEvent: "",
        eventsQueue: [],
        maxEventsInQueue: 7
    };

    const demoLayout = {
        demoLayoutTop: {
            backgroundColor: "#000000",
            opacity: "0.75",
            width: "100%",
            position: "fixed",
            top: "0",
            left: "0",
            zIndex: "9999"
        },
        demoLayoutCenter: {
            backgroundColor: "#000000",
            opacity: "0.75",
            width: "100%",
            position: "fixed",
            left: "0",
            zIndex: "9999",
            height: "5px",
            transform: "translateY(-50%)"
        },
        demoLayoutBottom: {
            backgroundColor: "#000000",
            opacity: "0.75",
            width: "100%",
            position: "fixed",
            top: "0",
            left: "0",
            zIndex: "9999"
        }
    };

    function Standout(element, i, options, isLast) {

        this._name = pluginName;
        this.i = i;
        this.idx = `standout_${this.i}`;
        this.clonedId = `${this.idx}_cloned`;
        this.isLastElement = isLast;
        this.options = this._getOptions(options);
        this.element = element;
        this.enabled = this.options.enabled;
        allStandout[i] = this;
        this._init();

    }

    // Avoid Plugin.prototype conflicts
    jQuery.extend(Standout.prototype, {

        _init() {

            const offset = this._isWindow() ? 0 : jQuery(this.options.window).offset().top;
            this._buildCache();
            this._initViewport(offset);
            this._initObj(offset);
            this._bindEvents();

            if (this.options.showDemoLayout && !demoInit) {

                demoInit = true;
                this._setDemoLayout(this.options, offset);

            }

            if (this.options.lightBoxEffect) {

                if (jQuery(`#${this.options.overlayId}`).length === 0) {

                    this._setOverlayLayout(this.options);

                }

                if (jQuery(`.${this.clonedId}`).length === 0) {

                    this._duplicateElement(this.element, this.clonedId, this.options);

                }

                if (this.options.compatibility && this.isLastElement && typeof this.options.dynamicContentListeners === "function") {

                    this.options._dynamicContentListeners();

                }

                this._registerLightboxEvents();

            }

        },

        _buildCache() {

            this.$element = this._getElement(this.i);

        },

        _bindEvents() {

            const obj = this;

            jQuery(obj.options.window).on("resize scroll", function() {

                obj._updateObj();

                if (obj._isInViewport()) {

                    obj._setCurrentEvent();
                    if ((!obj.options.oncePerEvent)
                    || (obj.options.onlyFirstTime && obj.options.oncePerEvent && !obj._checkIfEventAlreadyTriggered())
                    || (obj.options.oncePerEvent && !obj.options.onlyFirstTime && !obj._checkIfLastEvent())) {

                        // triggering the event passing along the current object
                        obj.$element.trigger(obj.$element.objProps.currentEvent, [obj]);

                    }
                    obj._setLastEvent();
                    obj._addEventToQueue();

                } else {

                    // Reset current event when element is not in viewport
                    obj._resetEvents();

                }

            });

        },

        _isInViewport() {

            return this.$element.objProps.elementBottomPosition > this.$element.objProps.extremeTop
                && this.$element.objProps.elementTopPosition < this.$element.objProps.extremeBottom
                && this.$element.objProps.initialized;

        },

        _setCurrentEvent() {

            const props = this.$element.objProps;
            let event = "U";
            if (props.lastViewportTop === 0) {

                props.lastViewportTop = props.viewportTop;

            }

            if (props.elementCenterPosition >= props.viewportTopLimit && props.elementCenterPosition <= props.viewportBottomLimit) {

                event = "C";

            } else {

                if (props.elementBottomPosition <= props.viewportTopLimit) {

                    event = "O";

                } else {

                    if (props.elementBottomPosition > props.viewportTopLimit && props.elementCenterPosition < props.viewportTopLimit) {

                        if (props.viewportTop < props.lastViewportTop) {

                            event = "ET";

                        } else {

                            event = "EXT";

                        }

                    } else if (props.elementTopPosition < props.viewportBottomLimit && props.elementCenterPosition > props.viewportBottomLimit) {

                        if (props.viewportTop > props.lastViewportTop) {

                            event = "EB";

                        } else {

                            event = "EXB";

                        }

                    }

                }

            }

            props.lastViewportTop = props.viewportTop;
            props.currentEvent = event;

        },

        _getCurrentEvent() {

            return this.$element.objProps.currentEvent;

        },

        _setLastEvent() {

            this.$element.objProps.lastEvent = this.$element.objProps.currentEvent;

        },

        _getLastEvent() {

            return this.$element.objProps.lastEvent;

        },

        _getViewportBottomLimit() {

            return this.$element.objProps.viewportBottomLimit;

        },

        _getViewportTopLimit() {

            return this.$element.objProps.viewportTopLimit;

        },

        _getElementCenterPosition() {

            return this.$element.objProps.elementCenterPosition;

        },

        _getElementHeight() {

            return this.$element.objProps.elementHeight;

        },

        _checkIfLastEvent() {

            return this._getLastEvent() === this._getCurrentEvent();

        },

        _checkIfEventAlreadyTriggered() {

            return this.$element.objProps.eventsQueue.includes(this._getCurrentEvent());

        },

        _addEventToQueue() {

            if (!this._checkIfEventAlreadyTriggered()) {

                this.$element.objProps.eventsQueue.push(this._getCurrentEvent());

            }

        },

        _resetEvents() {

            this.$element.objProps.lastEvent = "";
            this.$element.objProps.currentEvent = "";

        },

        _initViewport() {

            const isWindow = this._isWindow();
            const offset = isWindow ? 0 : jQuery(this.options.window).offset().top;
            this._updateViewport();
            this.$element.objProps.viewportHeight = jQuery(this.options.window).height();
            this.$element.objProps.viewportWidth = jQuery(this.options.window).width();
            this.$element.objProps.viewportBottom = this.$element.objProps.viewportTop + this.$element.objProps.viewportHeight;
            this.$element.objProps.viewportLeft = isWindow ? 0 : jQuery(this.options.window).offset().left;

            /* NEW METHOD */
            this.$element.objProps.centerLimit = ((this.$element.objProps.viewportHeight * this.options.top) + offset)
                                                    + (this.$element.objProps.viewportHeight * (1 - (this.options.top + this.options.bottom)) * 0.5);
            this.$element.objProps.distanceTop = this.$element.objProps.centerLimit - (this.$element.objProps.viewportHeight * (1 - (this.options.top + this.options.bottom)) * 0.5);
            this.$element.objProps.distanceBottom = this.$element.objProps.centerLimit + (this.$element.objProps.viewportHeight * (1 - (this.options.top + this.options.bottom)) * 0.5);

            this.$element.objProps.extremeTop = this.$element.objProps.distanceTop - (this.$element.objProps.viewportHeight * this.options.top);
            this.$element.objProps.extremeBottom = this.$element.objProps.distanceBottom + (this.$element.objProps.viewportHeight * this.options.bottom);

            this.$element.objProps.viewportTopLimit = this.$element.objProps.distanceTop;
            this.$element.objProps.viewportBottomLimit = this.$element.objProps.distanceBottom;
            this.$element.objProps.viewportCenterLimit = this.$element.objProps.centerLimit;

        },

        _updateViewport() {

            this.$element.objProps.viewportTop = jQuery(this.options.window).scrollTop();

        },

        _isWindow() {

            return this.options.window === window;

        },

        _initObj() {

            this.$element.objProps.originalTop = this.$element.offset().top;
            this.$element.objProps.originalLeft = this.$element.offset().left;
            this.$element.objProps.initialized = true;

        },

        _updateObj() {

            const props = this.$element.objProps;
            const isWindow = this._isWindow();
            this._updateViewport();
            props.initialized = true;
            props.display = this.$element.css("display");
            props.elementWidth = this.$element.outerWidth();
            props.elementHeight = this.$element.outerHeight();
            props.elementTop = this.$element.offset().top;
            props.elementCenter = props.elementTop + props.elementHeight / 2;
            props.elementBottom = props.elementTop + props.elementHeight;

            if (isWindow) {

                props.elementTopPosition = props.elementTop - props.viewportTop;
                props.elementCenterPosition = props.elementTopPosition + props.elementHeight / 2;
                props.elementBottomPosition = props.elementTopPosition + props.elementHeight;

            } else {

                props.elementTopPosition = props.elementTop;
                props.elementCenterPosition = props.elementCenter;
                props.elementBottomPosition = props.elementBottom;

            }

            // Check this for future offset more stable fix
            // https://stackoverflow.com/questions/30875646/jquery-offset-top-slightly-off-every-other-time

        },

        _calculateOpacityValue(limit) {

            const max = this._getElementHeight() / 2;
            const current = Math.abs(limit - this._getElementCenterPosition());
            const opacity = (max - current) / max;
            return opacity;

        },

        _getCurrentElementOpacity() {

            let opacity = 0;
            const status = this._getCurrentEvent();

            switch (status) {

                case "C":
                    opacity = 1;
                    break;
                case "EB":
                case "EXB":
                    opacity = this._calculateOpacityValue(this._getViewportBottomLimit());
                    break;
                case "ET":
                case "EXT":
                    opacity = this._calculateOpacityValue(this._getViewportTopLimit());
                    break;
                default:
                    break;

            }

            return opacity.toFixed(2);

        },

        _getPrevNextElementOpacity() {

            // Take the max value between current, next and prev elements in viewport
            let c = 0;
            let nxt = 0;
            let prev = 0;

            const nxtObj = this._getNextElement(this.i);
            const prvObj = this._getPrevElement(this.i);

            c = this._getCurrentElementOpacity();

            if (typeof nxtObj !== "undefined" && nxtObj.$element.hasOwnProperty("objProps")) {

                nxt = nxtObj._getCurrentElementOpacity();

            }

            if (typeof prvObj !== "undefined" && prvObj.$element.hasOwnProperty("objProps")) {

                prev = prvObj._getCurrentElementOpacity();

            }

            return Math.max(c, nxt, prev);

        },

        _registerLightboxEvents() {

            const obj = this;

            /* LIGHTBOX EVENTS */
            obj.$element.on("EB EXB ET EXT O  U C", function() {

                return obj._fading();

            });

        },

        _fading() {

            const props = this.$element.objProps;
            const overlayPercentage = this._getPrevNextElementOpacity();
            const displayOverlay = overlayPercentage <= 0 ? "none" : "block";
            const displayElement = this._getCurrentElementOpacity() <= 0 ? "none" : props.display;

            jQuery("#overlayStandout").css({
                "display": displayOverlay,
                "opacity": overlayPercentage < 0.75 ? overlayPercentage : 0.75
            });
            jQuery(`.${this.clonedId}`).css({
                "display": displayElement,
                "position": "absolute",
                "top": props.originalTop,
                "left": props.originalLeft,
                "z-index": "10000",
                "width": props.elementWidth,
                "height": props.elementHeight,
                "margin": "0",
                "opacity": this._getCurrentElementOpacity() < 1 ? this._getCurrentElementOpacity() : 1
            });

        },

        _getOptions(options) {

            return options.lightBoxEffect ? jQuery.extend({}, defaults, options, lightboxOptions) : jQuery.extend({}, defaults, options);

        },

        _getElement(i) {

            let obj = false;
            if (typeof allStandout[i] !== "undefined" && i > -1 && i < allStandout.length) {

                obj = jQuery(allStandout[i].element);
                obj.objProps = jQuery.extend({}, objProps);

            }
            return obj;

        },

        _getNextElement(i) {

            return allStandout[i + 1];

        },

        _getPrevElement(i) {

            return allStandout[i - 1];

        },

        _setDemoLayout(opt, offset) {

            const props = this.$element.objProps;
            demoLayout.demoLayoutTop.top = offset <= 0 ? offset : props.viewportTop + offset;
            demoLayout.demoLayoutBottom.top = this._getDemoLayoutBottomTopPosition(offset, props, opt);
            demoLayout.demoLayoutTop.width = props.viewportWidth;
            demoLayout.demoLayoutCenter.width = props.viewportWidth;
            demoLayout.demoLayoutBottom.width = props.viewportWidth;
            demoLayout.demoLayoutTop.left = props.viewportLeft;
            demoLayout.demoLayoutCenter.left = props.viewportLeft;
            demoLayout.demoLayoutBottom.left = props.viewportLeft;

            let height = props.viewportHeight * opt.top;
            let overlay = jQuery("<div />").css(demoLayout.demoLayoutTop).css("height", height).attr("id", "demoLayoutTop");
            jQuery(opt.body).append(overlay);
            const top = (demoLayout.demoLayoutTop.top + props.viewportHeight * opt.top) + (props.viewportHeight - (props.viewportHeight * opt.top + props.viewportHeight * opt.bottom)) / 2;
            overlay = jQuery("<div />").css(demoLayout.demoLayoutCenter).css("top", top);
            jQuery(opt.body).append(overlay);
            height = props.viewportHeight * opt.bottom;
            overlay = jQuery("<div />").css(demoLayout.demoLayoutBottom).css("height", height).attr("id", "demoLayoutBottom");
            jQuery(opt.body).append(overlay);

        },

        _getDemoLayoutBottomTopPosition(offset, props, opt) {

            let topPosition = (props.viewportBottom + offset) - props.viewportHeight * opt.bottom;

            if (offset <= 0) {

                topPosition = props.viewportHeight - props.viewportHeight * opt.bottom;

            }

            return topPosition;

        },

        _setOverlayLayout(opt) {

            const overlay = jQuery("<div />").css(opt.overlay).attr("id", opt.overlayId);
            jQuery(opt.body).append(overlay);

        },

        _duplicateElement(el, c, opt) {

            jQuery(el).clone(true, true).addClass(c).css("display", "none")
            .appendTo(opt.body);

        },
    });

    /* eslint-disable */
    jQuery.fn[pluginName] = function(options) {
    /* eslint-enable */

        if (options === undefined || typeof options === "object") {

            const allElements = this;
            const lastIdx = allElements.length - 1;
            return this.each(function(i) {

                // Only allow the plugin to be instantiated once,
                // so we check that the element has no plugin instantiation yet
                if (!jQuery.data(this, `plugin_${pluginName}`)) {

                    // if it has no instance, create a new one,
                    // pass options to our plugin constructor,
                    // and store the plugin instance
                    // in the elements jQuery data object.
                    jQuery.data(this, `plugin_${pluginName}`, new Standout(allElements[i], i, options, (i === lastIdx)));

                }

            });

        }

        // If the first parameter is a string and it doesn't start
        // with an underscore or "contains" the `init`-function,
        // treat this as a call to a public method.
        if (typeof options === "string" && options[0] !== "_" && options !== "init") {

            // Allow instances to be destroyed via the 'destroy' method
            if (options === "destroy") {

                jQuery.data(this, `plugin_${pluginName}`, null);

            }

        } else {

            console.log("No public method available");

        }

    };

}(jQuery, window, document));
